src/ranged.cpp

```diff
@@ -376,6 +382,52 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
     return attack;
 }
 
+double player::gun_effective_range( const item& gun, int aim, int penalty, unsigned chance, double accuracy ) const
+{
+    if( !gun.is_gun() || !gun.ammo_sufficient() ) {
+        return 0;
+    }
+
+    int driving = 0;
+    if( penalty < 0 ) {
+        // get current effective player recoil
+        penalty = std::max( recoil, 0 );
+        if( recoil < 0 ) {
+            debugmsg( "negative player recoil when calculating effective range" );
+        }
+
+        // get maximum penalty from driving
+        driving = std::max( driving_recoil, 0 );
+        if( driving_recoil < 0 ) {
+            debugmsg( "negative driving recoil when calculating effective range" );
+        }
+    }
+
+    // aim_per_time() returns improvement (in MOA) per 10 moves
+    for( int i = aim * 10; i != 0; --i ) {
+        int adj = aim_per_time( gun, penalty );
+        if( adj <= 0 ) {
+            break; // no further improvement is possible
+        }
+        penalty -= adj;
+    }
+
+    // calculate maximum potential dispersion
+    double dispersion = get_weapon_dispersion( &gun, false ) + penalty + driving;
+
+    // cap at min 1MOA as at zero dispersion would result in an infinite effective range
+    dispersion = std::max( dispersion, 1.0 );
```

Comment:
The `dispersion` still can become zero on the next line (given zero `chance`). I guess this check should be either moved down below or dropped entirely (it's not a big deal since `min( +x, +INF )` is always `+x`, but it must be guaranteed that `accuracy` is always greater than zero. Troubles otherwise).